sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/ZPageServer.java

```diff
@@ -0,0 +1,165 @@
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A collection of HTML pages to display stats and trace data and allow library configuration
+ * control.
+ *
+ * <p>Example usage with private {@link HttpServer}
+ *
+ * <pre>{@code
+ * public class Main {
+ *   public static void main(String[] args) throws Exception {
+ *     ZPageServer.startHttpServerAndRegisterAllPages(8000);
+ *     ... // do work
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Example usage with shared {@link HttpServer}
+ *
+ * <pre>{@code
+ * public class Main {
+ *   public static void main(String[] args) throws Exception {
+ *     HttpServer server = HttpServer.create(new InetSocketAddress(8000), 10);
+ *     ZPageServer.registerAllPagesToHttpServer(server);
+ *     server.start();
+ *     ... // do work
+ *   }
+ * }
+ * }</pre>
+ */
+@ThreadSafe
+public final class ZPageServer {
+  // The maximum number of queued incoming connections allowed on the HttpServer listening socket.
+  private static final int HTTPSERVER_BACKLOG = 5;
+  // Length of time to wait for the HttpServer to stop
+  private static final int HTTPSERVER_STOP_DELAY = 1;
+  // Tracez SpanProcessor and DataAggregator for constructing TracezZPageHandler
+  private static final TracezSpanProcessor tracezSpanProcessor =
+      TracezSpanProcessor.newBuilder().build();
+  private static final TracezDataAggregator tracezDataAggregator =
+      new TracezDataAggregator(tracezSpanProcessor);
+  // Handler for /tracez page
+  private static final ZPageHandler tracezZPageHandler =
+      TracezZPageHandler.create(tracezDataAggregator);
+
+  private static final Object mutex = new Object();
+  private static volatile boolean isTracezSpanProcesserAdded = false;
+
+  @GuardedBy("mutex")
+  @Nullable
+  private static HttpServer server;
+
+  /** Function that adds the {@link TracezSpanProcessor} to the {@link tracerSdkProvider}. */
+  private static void addTracezSpanProcessor() {
+    if (!isTracezSpanProcesserAdded) {
+      synchronized (mutex) {
+        if (isTracezSpanProcesserAdded) {
+          return;
+        }
+        TracerSdkProvider tracerProvider = OpenTelemetrySdk.getTracerProvider();
+        tracerProvider.addSpanProcessor(tracezSpanProcessor);
+        isTracezSpanProcesserAdded = true;
+      }
+    }
+  }
+
+  /**
+   * Returns a {@code ZPageHandler} for tracing debug. The page displays information about all
+   * running spans and all sampled spans based on latency and error.
+   *
+   * <p>It displays a summary table which contains one row for each span name and data about number
+   * of running and sampled spans.
+   *
+   * <p>Clicking on a cell in the table with a number that is greater than zero will display
+   * detailed information about that span.
+   *
+   * @return a {@code ZPageHandler} for tracing debug.
+   */
+  public static ZPageHandler getTracezZPageHandler() {
+    addTracezSpanProcessor();
```

Comment:
It's mysterious that a `get` method has a side effect like registering a span processor. We should at least document this in the javadoc.

Reply:
I agree. I think this method should be renamed to `createAndRegister...`

Reply:
How about `createTracezZPageHandler` ? It will be called to register the handler to the server

Reply:
Actually, I think `get...` makes some sense since `tracezZPageHandler` is already constructed/created, I did add documentation stating the side effect

Reply:
Should we really be registering it for the user? Can't they register it themselves, so there isn't this hidden side-effect on a method called 'get'? In Java, a method called 'get...' should not have side-effects, except in extraordinary circumstances.

Reply:
Does this even need to return anything? Could it just be "registerHandler" or something?

Reply:
This spanProcessor is not related to the user as it's only used by TraceZ zPage, we've discussed this in [design doc](https://docs.google.com/document/d/1jkrTYI8pQxt-AJRAJW5eUApOT1wha2jt-1QueNj8Brw/edit#heading=h.8orj11wueywb)

Reply:
I can also change the way of implementing these sets of functions to "registerTracezHandler" and register the handler to the server

Reply:
@anuraaga I've shared it with the email address on your GitHub account!

Reply:
Thanks I can see it! But I still have concerns - OpenTelemetry is an open project so all design docs must be open without needing access or even a Google account. The doc should persist even past internship periods, etc.

Can you at least make the current doc public to anyone? It's not ideal since Docs is often blocked in China but better than current private sharing. Even better may be to copy it into a otep so it lives in open source

https://github.com/open-telemetry/oteps

As an aside, go/bluedoc has great formatting for the content, but a version needs to be made with similar content that is markdown for work in open source in a GitHub-first manner. If you can help with that, it will make it a lot easier for the company to contribute to OSS.

Let me know if you need any more guidance on this.

Reply:
Thanks! I'll talk with my intern host regarding this!

Reply:
Due to recent updates, there's no longer a way to make a google doc public from a corp account. So we have two choices: share another doc from your personal account, or share the doc as a markdown file in the repo. I think the latter option makes more sense.

Reply:
Thanks for the clarification, yeah markdown seems to make a lot of sense.

@jkwatson I notice the oteps repo is mainly for spec additions so maybe not for language SDK design docs. Do you think any of these make sense?

- open-telemetry/opentelemetry-java/docs (current repo, docs subfolder to keep design docs for posterity)
- open-telemetry/opentelemetry-java-docs (java-specific docs repo)
- open-telemetry/docs/java (docs repo that can have language, or whatever, subfolders)

Reply:
Filed #1397 for this

Reply:
We will make adjustments to our design docs and make a public markdown version of them around the end of the term